File: 213-house-robber-ii/213-house-robber-ii.cpp
class Solution {
public:
    int solve(vector<int>& nums){
        int previ = nums[0],prev2i = 0;
        
        for(int i=1;i<nums.size();i++){
            auto ans = max(previ,prev2i+nums[i]);
            prev2i = previ;
            previ = ans;
        }
        
        return previ;
    }
    
    int rob(vector<int>& nums) {
        
        auto n = nums.size();
        if (n==1) return nums[0];
        auto lastElem = nums[n-1];
        nums.pop_back();
        auto firstElem = nums[0];
        auto withoutEnd = solve(nums);
        nums.push_back(lastElem);
        nums.erase(nums.begin());
        auto withoutBegin = solve(nums);
        nums.insert(nums.begin(),firstElem);
        return max(withoutEnd,withoutBegin);
    }
};

File: 213-house-robber-ii/213-house-robber-ii.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        if len(nums)==1:
            return nums[0]
        
        
        def solve(L,R):
            secondLast,last = 0,0
            cur = 0
            for i in range(L,R+1):
                cur = max(nums[i]+secondLast,last)
                secondLast = last
                last = cur
            return cur
        
        n = len(nums)
        return max(solve(1,n-1),solve(0,n-2))